ocaml/xapi/xapi_cluster_host.ml

```diff
@@ -79,41 +79,42 @@ let join_internal ~__context ~self =
 
       debug "Enabling clusterd and joining cluster_host %s" (Ref.string_of self);
       Xapi_clustering.Daemon.enable ~__context;
-      Db.Cluster_host.set_joined ~__context ~self ~value:true;
-      Db.Cluster_host.set_enabled ~__context ~self ~value:true;
-      Cluster_client.LocalClient.join (rpc ~__context) dbg cluster_token ip ip_list
+      match Cluster_client.LocalClient.join (rpc ~__context) dbg cluster_token ip ip_list with
+      | Result.Ok () ->
+        debug "Cluster join create was successful for cluster_host %s" (Ref.string_of self)
+        Db.Cluster_host.set_joined ~__context ~self ~value:true;
+        Db.Cluster_host.set_enabled ~__context ~self ~value:true;
+        debug "Cluster_host %s joined and enabled" (Ref.string_of self)
+      | Result.Error error ->
+        warn "Error occurred when joining cluster_host %s" (Ref.string_of self);
+        handle_error error
   )
 
 (* Enable cluster_host in client layer via clusterd *)
 let resync_host ~__context ~host =
   match (find_cluster_host ~__context ~host) with
   | None      -> () (* no clusters exist *)
   | Some self ->    (* cluster_host and cluster exist *)
-    if not (Db.Cluster_host.get_joined ~__context ~self) then begin
-       match join_internal ~__context ~self with
-      | Result.Ok () ->
-        debug "Cluster_host.create was successful; cluster_host: %s" (Ref.string_of self)
-      | Result.Error error ->
-        warn "Error occurred during Cluster_host.create";
-        handle_error error
-    end;
+    if not (Db.Cluster_host.get_joined ~__context ~self)
```

Comment:
This might look odd, but I think its better for ordering to keep it like this. The logical sequence is: join then later enable if not already enabled.